### PR TITLE
Add VolumeConfig cache based on hostname.

### PR DESCRIPTION
**Pull request by iamazeem**
Add VolumeConfig cache based on hostname.

Signed-off-by: Azeem Sajid <azeem.sajid@gmail.com>

### CODE REVIEW

File: cwm_worker_operator/domains_config.py [hunk 1]
```diff
@@ -233,6 +233,7 @@ def __init__(self, data, domains_config, request_hostname=None, is_data_from_cac
             request_data['__request_hostname'] = request_hostname
             self._last_update = request_data["__last_update"] = common.now().strftime("%Y%m%dT%H%M%S")
             domains_config.keys.volume_config.set(request_worker_id, json.dumps(request_data))
+            domains_config.keys.volume_config.set(request_hostname, request_worker_id)
```
Comment (OriHoch) [inline cwm_worker_operator/domains_config.py:236]: it's a good idea to do it this way, but it should be in a new key, otherwise it may cause unexpected problems and complicates the logic. Also, it should use the instance id (`self.id` - because request_worker_id may not always be set.

e.g. `domains_config.keys.volume_config_hostname_worker_id.set(request_hostname, self.id)`

when you add a new key, be sure to also delete it and handle all the lifecycle of key, you should do a global search for the volume_config key and make sure all of the same operations are handled for the new key as well

another issue here is that the code depends on request_worker_id being set, but this key should be set regardless of that - under an new "if" statement

also, the condition `(request_hostname is not None and data.get('__request_hostname') != request_hostname)` - is not relevant for storing this key

File: cwm_worker_operator/domains_config.py [hunk 2]
```diff
@@ -409,9 +410,12 @@ def get_cwm_api_volume_config(self, metrics=None, force_update=False, hostname=N
...
             raise Exception("either hostname or worker_id param is required")
         start_time = common.now()
+        val = None
         if force_update or not worker_id:
-            # TODO: support cache for getting volume config based on hostname
-            val = None
+            cached_worker_id = self.keys.volume_config.get(hostname)
+            if cached_worker_id:
+                cached_worker_id = cached_worker_id.decode()
+                val = json.loads(self.keys.volume_config.get(cached_worker_id))
```
Comment (OriHoch) [inline cwm_worker_operator/domains_config.py:418]: the hostname and worker_id cache should not be dependant on each other

if worker_id is set - get the volume config from worker_id
if hostname is set - get the volume config from hostname
if force_update is true - don't use the cache

also, you should not handle decoding of internal Redis representation here. You should use a separate key of class `DomainsConfigKeyPrefixInt` which will handle this and return an int

File: cwm_worker_operator/domains_config.py [hunk 3]
```diff
@@ -448,11 +452,12 @@ def get_cwm_api_volume_config(self, metrics=None, force_update=False, hostname=N
                     metrics.cwm_api_volume_config_error_from_api(worker_id or 'missing', start_time)
             return VolumeConfig(volume_config, self, request_hostname=hostname, is_data_from_cache=False, request_worker_id=worker_id)
         else:
+            volume_config = json.loads(val)
+            if not worker_id:
+                worker_id = volume_config['instanceId']
```
Comment (OriHoch) [inline cwm_worker_operator/domains_config.py:457]: you should not parse the volume config here, the parsing logic may change and should be only handled internally in the VolumeConfig class
instead, you should set volume config to a variable and then use volume_config.id

**Review by OriHoch — CHANGES_REQUESTED** [review]
see comments, I think that some of the changes were not backwards compatible and changed some of the logic - so it may be the reason for failed unit tests

**Comment by iamazeem** [comment]
@OriHoch 

With the addition of the new key (`volume_config_hostname_worker_id`), the test where the values from all the pools are being compared will fail.

**Example**

Assertion:

```python
    assert domains_config._get_all_redis_pools_values(blank_keys=[volume_config_key]) == {
        volume_config_key: "",
        hostname_error_key: 'INVALID_VOLUME_ZONE'
    }
```

Error (`test_initialize_invalid_volume_zone`):

```text
________________________________________________________________________________________________ test_initialize_invalid_volume_zone ________________________________________________________________________________________________

domains_config = <tests.mocks.domains_config.MockDomainsConfig object at 0x7f5f4c613520>, initializer_metrics = <tests.mocks.metrics.MockInitializerMetrics object at 0x7f5f4c613e20>

    def test_initialize_invalid_volume_zone(domains_config, initializer_metrics):
        worker_id, hostname = 'worker1', 'invalid-zone.com'
        domains_config.keys.hostname_initialize.set(hostname, '')
        volume_config_key = domains_config.keys.volume_config._(worker_id)
        hostname_error_key = domains_config.keys.hostname_error._(hostname)
        # set mock volume config in api with invalid zone
        domains_config._cwm_api_volume_configs['hostname:{}'.format(hostname)] = {
            'instanceId': worker_id, 'zone': 'INVALID', 'minio_extra_configs': {'hostnames': [{'hostname': hostname}]}
        }
        initializer.run_single_iteration(domains_config, initializer_metrics)
        print(">>> volume_config_key: {}".format(volume_config_key))
        values = domains_config._get_all_redis_pools_values(blank_keys=[volume_config_key])
        print("\n>>> values: {}\n".format(values))
>       assert domains_config._get_all_redis_pools_values(blank_keys=[volume_config_key]) == {
            volume_config_key: "",
            hostname_error_key: 'INVALID_VOLUME_ZONE'
        }
E       AssertionError: assert {'hostname:error:invalid-zone.com': 'INVALID_VOLUME_ZONE',\n 'worker:volume:config:hostname_worker_id:invalid-zone.com': 'worker1',\n 'worker:volume:config:worker1': ''} == {'hostname:error:invalid-zone.com': 'INVALID_VOLUME_ZONE',\n 'worker:volume:config:worker1': ''}
E         Common items:
E         {'hostname:error:invalid-zone.com': 'INVALID_VOLUME_ZONE',
E          'worker:volume:config:worker1': ''}
E         Left contains 1 more item:
E         {'worker:volume:config:hostname_worker_id:invalid-zone.com': 'worker1'}
E         Full diff:
E           {
E            'hostname:error:invalid-zone.com': 'INVALID_VOLUME_ZONE',
E         +  'worker:volume:config:hostname_worker_id:invalid-zone.com': 'worker1',
E            'worker:volume:config:worker1': '',
E           }

tests/test_initializer.py:59: AssertionError
====================================================================================================== short test summary info ======================================================================================================
FAILED tests/test_initializer.py::test_initialize_invalid_volume_zone - AssertionError: assert {'hostname:error:invalid-zone.com': 'INVALID_VOLUME_ZONE',\n 'worker:volume:config:hostname_worker_id:invalid-zone.com': 'worker1',...
================================================================================================= 1 failed, 68 deselected in 0.60s ==================================================================================================
```

Additional KV pair:

```text
'worker:volume:config:hostname_worker_id:invalid-zone.com': 'worker1'
```

**Suggested solution (tested):**

Only the required values would need to tested e.g.:

```python
    values = domains_config._get_all_redis_pools_values(blank_keys=[volume_config_key])
    assert values[volume_config_key] == ''
    assert values[hostname_error_key] == 'INVALID_VOLUME_ZONE'
```

Please let me know if this approach is good to go.

**Comment by OriHoch** [comment]
the test needs to be fixed - the additional key needs to be added to the expected result

File: cwm_worker_operator/domains_config.py [hunk 4]
```diff
@@ -448,11 +455,10 @@ def get_cwm_api_volume_config(self, metrics=None, force_update=False, hostname=N
                     metrics.cwm_api_volume_config_error_from_api(worker_id or 'missing', start_time)
             return VolumeConfig(volume_config, self, request_hostname=hostname, is_data_from_cache=False, request_worker_id=worker_id)
         else:
-            if metrics:
-                # success from cache is only possible when we got a worker_id
-                # TODO: add support for cache based on hostname
+            if metrics and worker_id:
                 metrics.cwm_api_volume_config_success_from_cache(worker_id, start_time)
```
Comment (OriHoch) [inline cwm_worker_operator/domains_config.py:459]: metrics report should be sent even if worker_id was not provided

code should be like this:

```
volume_config = VolumeConfig(json.loads(val), self, request_hostname=hostname, is_data_from_cache=True, request_worker_id=worker_id)
if metrics and volume_config.id:
  metrics.cwm_api_volume_config_success_from_cache(worker_id, start_time)
return volume_config
```

File: cwm_worker_operator/domains_config.py [hunk 5]
```diff
@@ -582,6 +591,9 @@ def del_worker_keys(self, worker_id, with_error=True, with_volume_config=True, w
                     r.delete(*keys)
         if with_volume_config:
             self.keys.volume_config.delete(worker_id)
+            for hostname in hostnames:
+                self.keys.volume_config_hostname_worker_id.delete(hostname)
```
Comment (OriHoch) [inline cwm_worker_operator/domains_config.py:595]: deletion of hostname keys should be handled in del_worker_hostname_keys

File: tests/test_domains_config.py [hunk 6]
```diff
@@ -273,13 +273,13 @@ def test_del_worker_keys(domains_config):
             key.set(namespace_name, val)
         else:
             raise Exception("Invalid keys_summary_param: {}".format(keys_summary_param))
-    domains_config.del_worker_keys(worker_id)
+    domains_config.del_worker_keys(worker_id, with_volume_config=True)
```
Comment (OriHoch) [inline tests/test_domains_config.py:276]: why was this change needed? I don't think any changes are needed in this test

**Review by OriHoch — CHANGES_REQUESTED** [review]
looks good, see comments
please also add a test method which tests the new functionality

File: tests/test_initializer.py [hunk 7]
```diff
@@ -111,6 +115,38 @@ def test_initialize_valid_domain(domains_config, initializer_metrics):
     assert [','.join(o['labels']) for o in initializer_metrics.observations] == [',success', ',initialized', ',success', ',initialized']
 
 
+def test_volume_config_hostname_worker_id_cache(domains_config, initializer_metrics):
```
Comment (OriHoch) [inline tests/test_initializer.py:118]: the change is in domains_config get_cwm_api_volume_config, so the test should test that function and be in test_domains_config

**Review by OriHoch — CHANGES_REQUESTED** [review]

File: tests/test_domains_config.py [hunk 8]
```diff
@@ -168,6 +168,21 @@ def test_get_volume_config_error(domains_config):
...
+def test_volume_config_hostname_worker_id_cache(domains_config):
+    dc = domains_config
+    worker_id = 'worker1'
+    hostname = 'example007.com'
+    dc._cwm_api_volume_configs['hostname:{}'.format(hostname)] = {
+        'instanceId': worker_id, 'minio_extra_configs': {'hostnames': [{'hostname': hostname}]}
+    }
+    volume_config = dc.get_cwm_api_volume_config(hostname=hostname)
+    assert volume_config.id == worker_id
+    assert domains_config.keys.volume_config_hostname_worker_id.get(hostname).decode() == worker_id
```
Comment (OriHoch) [inline tests/test_domains_config.py:180]: you need to verify that the volume config is received from cache, currently this test only verifies that the redis key is set

you ca verify that by deleting or setting an invalid value in `dc._cwm_api_volume_configs['hostname:{}'.format(hostname)]` then making another call to `get_cwm_api_volume_config` - but because the value is in cache it will still return the correct value from the cache

Comment (iamazeem) [reply tests/test_domains_config.py:180]: Currently, the `worker_id` is cached against `hostname`, not the complete `VolumeConfig` object.
The cached `worker_id` will be returned in this case if accessed from the cache.
Should I change it and cache the `VolumeConfig` to test the reset/delete scenario?

For example:

**domain_config.py**

```python
        if domains_config and request_hostname and self.id is not None:
            domains_config.keys.volume_config_hostname_worker_id.set(request_hostname, json.dumps(request_data))
```

**tests/test_domains_config.py**

```python
def test_volume_config_hostname_worker_id_cache(domains_config):
    dc = domains_config
    worker_id = 'worker1'
    hostname = 'example007.com'
    dc._cwm_api_volume_configs['hostname:{}'.format(hostname)] = {
        'instanceId': worker_id, 'minio_extra_configs': {'hostnames': [{'hostname': hostname}]}
    }
    volume_config_1 = dc.get_cwm_api_volume_config(hostname=hostname)
    assert domains_config.keys.volume_config_hostname_worker_id.get(hostname)

    # reset volume config and try to get from cache
    dc._cwm_api_volume_configs['hostname:{}'.format(hostname)] = {}
    volume_config_2 = dc.get_cwm_api_volume_config(hostname=hostname)
    assert domains_config.keys.volume_config_hostname_worker_id.get(hostname)

    assert volume_config_1 != volume_config_2
    assert volume_config_1.id == volume_config_2.id
    assert volume_config_1.hostnames == volume_config_2.hostnames

    dc.del_worker_hostname_keys(hostname)
    assert domains_config.keys.volume_config_hostname_worker_id.get(hostname) == None
```

File: tests/test_domains_config.py [hunk 9]
```diff
@@ -168,6 +168,27 @@ def test_get_volume_config_error(domains_config):
...
+
+    # empty volume configs and try to get worker_id from cache
+    volume_config = dc._cwm_api_volume_configs['hostname:{}'.format(hostname)] = {}
```
Comment (OriHoch) [inline tests/test_domains_config.py:183]: all the code from this line and until the end of the function doesn't call the actual function, it just tests internal redis / dicts which are irelevant for this test

you need to set `dc._cwm_api_volume_configs['hostname:{}'.format(hostname)] = {}`
then make another call to get_cwm_api_volume_config which should return the relevant volume config from the cache

File: tests/test_domains_config.py (hunk 8 above)
Comment (OriHoch) [reply tests/test_domains_config.py:180]: no, you should only cache the worker_id, the implementation seems fine, just need to write a test for it

File: tests/test_domains_config.py (hunk 9 above)
Comment (OriHoch) [reply tests/test_domains_config.py:183]: unit tests needs to test the relevant functionality which your PR is implementing, striving for 100% coverage of all relevant code paths

**Review by OriHoch — APPROVED** [review]
